packages/emitter-framework/src/typescript/components/union-expression.tsx

```diff
@@ -32,10 +52,73 @@ export function UnionExpression({ type, children }: UnionExpressionProps) {
   if (children || (Array.isArray(children) && children.length)) {
     return (
       <>
-        {variants} {` | ${children}`}
+        {variants} {`| ${children}`}
       </>
     );
   }
 
   return variants;
 }
+
+interface ObjectEnvelopeProps {
+  type: UnionVariant;
+  discriminatorPropertyName: string;
+  envelopePropertyName: string;
+}
+
+/**
+ * Renders a discriminated union with "object" envelope style
+ * where model properties are nested inside an envelope
+ */
+function ObjectEnvelope(props: ObjectEnvelopeProps) {
+  const { $ } = useTsp();
+
+  const envelope = $.model.create({
+    properties: {
+      [props.discriminatorPropertyName]: $.modelProperty.create({
+        name: props.discriminatorPropertyName,
+        type: $.literal.createString(props.type.name as string),
+      }),
+      [props.envelopePropertyName]: $.modelProperty.create({
+        name: props.envelopePropertyName,
+        type: props.type.type,
+      }),
+    },
+  });
+
+  return <TypeExpression type={envelope} />;
+}
+
+interface NoneEnvelopeProps {
+  type: UnionVariant;
+  discriminatorPropertyName: string;
+}
+
+/**
+ * Renders a discriminated union with "none" envelope style
+ * where discriminator property sits alongside model properties
+ */
+function NoneEnvelope(props: NoneEnvelopeProps) {
+  const { $ } = useTsp();
+
+  compilerAssert(
+    $.model.is(props.type.type),
+    "Expected all union variants to be models when using a discriminated union with no envelope",
+  );
+
+  // This is an anonymous type, so we render its properties along the discriminator property
+  if ($.model.isExpresion(props.type.type)) {
+    const model = $.model.create({
+      properties: {
+        [props.discriminatorPropertyName]: $.modelProperty.create({
+          name: props.discriminatorPropertyName,
+          type: $.literal.createString(props.type.name as string),
+        }),
+        ...Object.fromEntries(props.type.type.properties),
+      },
+    });
+    return <TypeExpression type={model} />;
+  }
+
+  return ay.code`{${props.discriminatorPropertyName}: "${String(props.type.name)}"} & ${efRefkey(props.type.type)}`;
```

Comment:
Still need to update this code to use proper templating with Alloy types, commit coming shortly